### DISCUSSION TITLE
Extracting entity types for use in the frontend

### DISCUSSION THREAD

**Discussion by tronikelis** [mikro-orm/mikro-orm#3546] [Q&A]
Hello,

What would be the best way to extract types from the entity classes to use in the frontend for example.

This is my best current attempt

```ts
import type { EntityDTO, Loaded } from "@mikro-orm/core";
import type { User } from "@server/db/entities";

interface UserJSON
    extends Omit<EntityDTO<Loaded<User, "UserRole">>, "Sessions" | "ExamSimAttempts"> {
    isRoot: boolean;
}
```

This is not ideal because it adds all associated entities:

![image](https://user-images.githubusercontent.com/56039679/192753575-77fd7500-8095-4f60-86f9-9cb21ced1d7c.png)

I want to specify that the `User` entity is populated **only** with the `UserRole` entity
And that `UserRole` entity ideally to be without any additional entities

Thanks

**Comment by rcfox** [comment]
I've just started thinking about this myself. I think ideally, the frontend should have no references to `@mikro-orm/core`, or at most type-only imports.

Perhaps a hypothetical `@mikro-orm/frontend` could provide no-op functions for the various decorators used in entity files, allowing us to use the entity files directly.

Alternatively, the entity generator could generate files for use by the frontend that would be structurally equivalent to the entities it generates now. (I'm not sure if this would be difficult to do with relationships.)

I don't know enough about the internals of MikroORM to know if either of these approaches would be feasible.

It would also be great if typeguard functions were provided to validate JSON.

**Reply by tronikelis** [reply]
Creating a new package just for the types seems like a stretch
I think the ideal simple solution would be a new generic type that you pass in your entity and comes out the jsonified output, so dates get converted to strings etc.

My current workaround is this:

Define all your possible entity names as a type
import type all your entities
export a new type that doesn't have any relations
Specify the attributes that you want to override (convert Date into string as seen in the example below) 
join your relations in another type (in another file preferably)

```ts
import type {
    Permission as IPermission,
    User as IUser,
} from "@server/db/entities";

interface CreatedUpdatedAt {
    createdAt: string;
    updatedAt: string;
}

type Overwrite<T, U> = Omit<T, keyof U> & U;

type AllEntities =
    | "Permission"
    | "Permissions"
    | "Users"
    | "User"

type CleanEntity<T> = Overwrite<Omit<T, AllEntities>, CreatedUpdatedAt>;

export type Permission = CleanEntity<IPermission>;
export type User = CleanEntity<IUser>;
```

How I join the relations in the frontend code

```ts
import { Certification, Course, Lesson } from "../types";

interface CourseJSON extends Course {
    Certification: Certification;
    Lessons: Lesson[];
}
```
